compiler/damlc/lib/DA/Cli/Damlc/Command/MultiIde/Types.hs

```diff
@@ -0,0 +1,182 @@
+-- Copyright (c) 2023 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE ApplicativeDo       #-}
+{-# LANGUAGE RankNTypes       #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE DisambiguateRecordFields #-}
+{-# LANGUAGE GADTs #-}
+
+module DA.Cli.Damlc.Command.MultiIde.Types (
+  module DA.Cli.Damlc.Command.MultiIde.Types
+) where
+
+import Control.Concurrent.Async (Async)
+import Control.Concurrent.STM.TChan
+import Control.Concurrent.STM.TVar
+import Control.Concurrent.STM.TMVar
+import Control.Concurrent.MVar
+import Control.Monad.STM
+import qualified Data.ByteString.Lazy as BSL
+import qualified Data.IxMap as IM
+import qualified Data.Map as Map
+import qualified Data.Text as T
+import qualified Language.LSP.Types as LSP
+import System.IO.Extra
+import System.Process (Pid)
+import System.Process.Typed (Process)
+
+data TrackedMethod (m :: LSP.Method from 'LSP.Request) where
+  TrackedSingleMethodFromClient
+    :: forall (m :: LSP.Method 'LSP.FromClient 'LSP.Request)
+    .  LSP.SMethod m
+    -> TrackedMethod m
+  TrackedSingleMethodFromServer
+    :: forall (m :: LSP.Method 'LSP.FromServer 'LSP.Request)
+    .  LSP.SMethod m
+    -> FilePath -- Also store the IDE that sent the request
+    -> TrackedMethod m
+  TrackedAllMethod :: forall (m :: LSP.Method 'LSP.FromClient 'LSP.Request).
+    { tamMethod :: LSP.SMethod m
+        -- ^ The method of the initial request
+    , tamLspId :: LSP.LspId m
+    , tamCombiner :: ResponseCombiner m
+        -- ^ How to combine the results from each IDE
+    , tamRemainingResponseIDERoots :: [FilePath]
+        -- ^ The IDES that have not yet replied to this message
+    , tamResponses :: [(FilePath, Either LSP.ResponseError (LSP.ResponseResult m))]
+    } -> TrackedMethod m
+
+tmMethod
+  :: forall (from :: LSP.From) (m :: LSP.Method from 'LSP.Request)
+  .  TrackedMethod m
+  -> LSP.SMethod m
+tmMethod (TrackedSingleMethodFromClient m) = m
+tmMethod (TrackedSingleMethodFromServer m _) = m
+tmMethod (TrackedAllMethod {tamMethod}) = tamMethod
+
+type MethodTracker (from :: LSP.From) = IM.IxMap @(LSP.Method from 'LSP.Request) LSP.LspId TrackedMethod
+type MethodTrackerVar (from :: LSP.From) = TVar (MethodTracker from)
+
+data SubIDE = SubIDE
+  { ideInhandleAsync :: Async ()
+  , ideInHandle :: Handle
+  , ideInHandleChannel :: TChan BSL.ByteString
+  , ideOutHandleAsync :: Async ()
+    -- ^ For sending messages to that SubIDE
+  , ideProcess :: Process Handle Handle ()
+  , ideProcessID :: Pid 
+  , ideHomeDirectory :: FilePath
+  , ideMessageIdPrefix :: T.Text
+    -- ^ Some unique string used to prefix message ids created by the SubIDE, to avoid collisions with other SubIDEs
+    -- We use the stringified process ID
+  , ideActive :: Bool
```

Comment:
Maybe a generic function that only gets active subIDEs

Reply:
Theres not really much I can do here actually, given we're not using the `with` approach yet, because I need the original set to put back into the var.
Best I can do rn is an `onlyActive :: SubIDEs -> SubIDEs` that filters out on the value itself.
We might be able to improve it later.

Reply:
Theres not really much I can do here actually, given we're not using the `with` approach yet, because I need the original set to put back into the var.
Best I can do rn is an `onlyActive :: SubIDEs -> SubIDEs` that filters out on the value itself.
We might be able to improve it later.